File: 1865. Finding Pairs With a Certain Sum.cpp
class FindSumPairs {
public:
    vector<int> nums;
    unordered_map<int, int> storeCount, storeCount2;
    FindSumPairs(vector<int>& nums1, vector<int>& nums2) 
    {
        nums = nums2;
        for(auto &val : nums1) storeCount[val] += 1;
        for(auto &val : nums2) storeCount2[val] += 1;
    }
    
    void add(int index, int val) 
    {
        storeCount2[nums[index]] -= 1;
        nums[index] += val;
        storeCount2[nums[index]] += 1;
    }
    
    int count(int tot) 
    {
        int ans = 0;
        for(auto [val, cnt] : storeCount2)
            ans += storeCount[tot - val] * cnt;
        return ans;
    }
};

File: Design/1865. Finding Pairs With a Certain Sum.py
class FindSumPairs:
    def __init__(self, nums1: List[int], nums2: List[int]):
        self.nums1 = nums1
        self.nums2 = nums2
        self.nums2_count = Counter(nums2)

    def add(self, index: int, val: int) -> None:
        self.nums2_count[self.nums2[index]] -= 1
        if self.nums2_count[self.nums2[index]] <= 0:
            self.nums2_count.pop(self.nums2[index])

        self.nums2[index] += val
        self.nums2_count[self.nums2[index]] += 1
        
    def count(self, tot: int) -> int:
        count = 0
        for val in self.nums1:
            count += self.nums2_count[tot - val]
        return count